From 62af1dc5d959898d02a2abfc435be04a864cf6bb Mon Sep 17 00:00:00 2001
From: Amlana Bhuyan <amlanab@ami.com>
Date: Fri, 2 Aug 2024 19:44:35 +0530
Subject: [PATCH] OT-4090 Addition of symlink-check

Issue:
 Scenario-1: After clicking the start and stop buttons multiple
 times in succession to mount, clicking start will automatically
 stop redirection.

 Scenario-2: After reproducing the Scenario-1 , close the web window.
 then reopen it and log-in,there will be no option for redirection.

Root-cause:
 Scenario-1 and Scenario-2  occurs because of service crashing caused
 by filesystem I/O error.

 In case service crashes while media redirection is in progress the old
 symlink doesn't get cleaned up, as a result the restarted service
 will throw error while creating a new symlink causing the service
 to crash again.

Resolution:
 Delay in between Stop/Start of consecutive media redirection.

 Symlink-check addition.So that old symlink can be removed before
 attempting to create a new one, preventing multiple time service
 crashing.

Signed-off-by: Amlana Bhuyan <amlanab@ami.com>
---
 src/system.hpp | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/system.hpp b/src/system.hpp
index c326c83..a7c091c 100644
--- a/src/system.hpp
+++ b/src/system.hpp
@@ -799,6 +799,13 @@ struct UsbGadget
                 echoToFile(configStringsDir / "configuration", "config 1");
                 fs::create_directories(funcMassStorageDir);
                 fs::create_directories(funcMassStorageDir / "lun.0");
+                /*symlink-check*/
+                if (fs::is_symlink(massStorageDir))
+                {
+                    LogMsg(Logger::Info,
+                           "Removing old symlink: ",massStorageDir.c_str());
+                    fs::remove(massStorageDir);
+                }
                 fs::create_directory_symlink(funcMassStorageDir,
                                              massStorageDir);
                 echoToFile(funcMassStorageDir / "lun.0/removable", "1");
--
2.25.1

